src/dmd/dclass.d

```diff
@@ -992,6 +992,23 @@ extern (C++) class ClassDeclaration : AggregateDeclaration
     {
         v.visit(this);
     }
+
+    /****************************************
+     * Gets an array of all implemented interfaces of this class. This includes
+     * all interfaces implemented by parent classes, all interfaces implemented
+     * by this class, and all interfaces which are extended by other interfaces.
+     */
+    final void allInterfaces(ref BaseClasses bases)
+    {
+        if (baseClass is null)
+            return;
+        baseClass.allInterfaces(bases);
+        foreach (base; interfaces)
+        {
+            bases.push(base);
+            base.sym.allInterfaces(bases);
+        }
+    }
```

Comment:
Looking at the implementation, this appears to be a form of "accumulate". It appends classes to `bases`.

Reply:
It was indeed. I have changed it to a function which takes a callback instead. I'm not entirely certain how I'd have to write it so that it can still be used with `foreach`, especially given that recursion is involved. I guess I could return an anonymous struct with an `opApply`, which can then call itself recursively while passing a delegate.

Reply:
Something like this should work, no? (spec: https://dlang.org/spec/statement.html#foreach_over_struct_and_classes)

```d
extern(D) final int opApply(scope int delegate(BaseClass*) callback)
{
        if (baseClass is null)
            return 0;

        int res = baseClass.opApply(callback);
        if (res) return res;

        foreach (base; interfaces)
        {
            res = callback(base);
            if (res) return res;

            res = base.sym.opApply(callback);
            if (res) return res;
        }
        return 0;
}
```

But the current implementation is good enough IMO, given that the doesn't need to `break`

Reply:
Other places in the frontend just call it `_foreach`.